xapian-letor/api/feature_internal.cc

```diff
@@ -0,0 +1,181 @@
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/feature.h"
+#include "feature_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+Database
+Feature::Internal::get_database() const
+{
+    LOGCALL(API, Database, "Feature::Internal::get_database", NO_ARGS);
+    return feature_db;
+}
+
+Query
+Feature::Internal::get_query() const
+{
+    LOGCALL(API, Query, "Feature::Internal::get_query", NO_ARGS);
+    return feature_query;
+}
+
+Document
+Feature::Internal::get_doc() const
+{
+    LOGCALL(API, Document, "Feature::Internal::get_doc", NO_ARGS);
+    return feature_doc;
+}
+
+bool
+Feature::Internal::get_termfreq(const std::string & term,
+			double & termfreq_) const
+{
+    LOGCALL(API, bool, "Feature::Internal::get_termfreq", term | termfreq_);
+    auto termfreq_iterator = termfreq.find(term);
+    if (termfreq_iterator != termfreq.end()) {
+	termfreq_ = (double)termfreq_iterator->second;
+	return true;
+    }
+    return false;
+}
+
+bool
+Feature::Internal::get_inverse_doc_freq(const std::string & term,
+			double & inverse_doc_freq_) const
+{
+    LOGCALL(API, bool, "Feature::Internal::get_inverse_doc_freq", term | inverse_doc_freq_);
+    auto inverse_doc_freq_iterator = inverse_doc_freq.find(term);
+    if (inverse_doc_freq_iterator != inverse_doc_freq.end()) {
+	inverse_doc_freq_ = (double)inverse_doc_freq_iterator->second;
+	return true;
+    }
+    return false;
+}
+
+bool
+Feature::Internal::get_doc_length(const std::string & term,
+			double & doc_length_) const
+{
+    LOGCALL(API, bool, "Feature::Internal::get_doc_length", term | doc_length_);
+    auto doc_length_iterator = doc_length.find(term);
+    if (doc_length_iterator != doc_length.end()) {
+	doc_length_ = (double)doc_length_iterator->second;
+	return true;
+    }
+    return false;
+}
+
+bool
+Feature::Internal::get_collection_length(const std::string & term,
+			double & collection_length_)
+{
+    LOGCALL(API, bool, "Feature::Internal::get_collection_length", term | collection_length_);
+    auto collection_length_iterator = collection_length.find(term);
+    if (collection_length_iterator != collection_length.end()) {
+	collection_length_ = (double)collection_length_iterator->second;
```

Comment:
And this shouldn't be `double` either - almost none of these stats should be...

This one is `Xapian::termcount` in the map currently, though maybe `Xapian::totallength` is a more appropriate type as `Xapian::termcount` could be overflowed by the total document length of a large database.

Reply:
fixed in a fixup commit.